### PR TITLE
feat!: dbtp 928 add cdn endpoint module

### PR DESCRIPTION
**Pull request by ejayesh**
This change introduces a new module called CDN, this will enable the deployment of a CloudFront endpoint.  
There are breaking changes in extensions.yml to facilitate this change.

Upgrade path: in extensions.yml update type: alb "cdn_domains_list" variable to include applications internal prefix.  See README.md/CDN for example.

### CODE REVIEW

File: .checkov.baseline [hunk 1]
```diff
@@ -1,5 +1,19 @@
...
+        {
+            "file": "/cdn/main.tf",
+            "findings": [
+                {
+                    "resource": "module.extensions-staging.module.cdn.aws_cloudfront_distribution.standard",
+                    "check_ids": [
+                        "CKV_AWS_305",
+                        "CKV_AWS_310",
+                        "CKV2_AWS_32",
+                        "CKV2_AWS_47"
```
Comment (WillGibson) [inline .checkov.baseline:12]: This is not a good way to address Checkov's findings.

We should not be adding new things to the Checkov baseline file.

It is only intended to allow easy adoption of checkov by temporarily silencing warnings about existing misconfigurations whilst still alerting about new ones.

The existing misconfigurations should be addressed as time allows and the baseline file will end up empty. We have a ticket for that.

If it alerts about something we genuinely need Checkov to ignore, the checks can be skipped in context in the files which should be accompanied by a comment explaining why it is skipped for our future selves benefit.

Comment (ejayesh) [reply .checkov.baseline:12]: None of these are relavant to our configuration of CloudFront. I didnt know how else to skip the check

Comment (ejayesh) [reply .checkov.baseline]: moving these

**Comment by WillGibson** [comment]
Title should start `feat!:` to indicate breaking change.

Pull request description should include Upgrade Path, for easy copying into the release notes.

File: .checkov.baseline [hunk 2]
```diff
@@ -27,7 +27,8 @@
                 {
                     "resource": "module.extensions-staging.module.alb.aws_security_group.alb-security-group",
                     "check_ids": [
-                        "CKV_AWS_23"
+                        "CKV_AWS_23",
+                        "CKV2_AWS_5"
```
Comment (WillGibson) [inline .checkov.baseline:31]: Why is `CKV2_AWS_5` still here?

File: README.md [hunk 3]
```diff
@@ -92,19 +92,52 @@ The R53 domains for non-production and production are stored in different AWS ac
...
+      enable_cdn_record: false
+      enable_logging: true
     prod:
-      domain: {my-application.great.gov.uk: "great.gov.uk"} 
+      cdn_domains_list: {my-application.great.gov.uk: "great.gov.uk"} 
 ```
 
+Each item in the `cdn_domain_list` must include:
+- Key: The endpoint name <myapp.mysite.com>
+- Values: application internal prefix and base domain <"internal", "mysite.com">
```
Comment (WillGibson) [inline README.md:130]:
```suggestion
```

File: README.md [hunk 4]
```diff
@@ -92,19 +92,52 @@ The R53 domains for non-production and production are stored in different AWS ac
...
+    dev: 
+      additional_address_list:
+        - internal.my-web-service-2
+```
+
+## CDN
+
+This module will create the CloudFront (CDN) endpoints for the application if enabled.
+
+Additional domains (`cdn_domains_list`) are the domain names that will be configured in CloudFront. In the map the key is the fully qualified domain name and the value is the application's base domain (the application's Route 53 zone).
```
Comment (WillGibson) [inline README.md:108]: It's confusing to us "Additional domains" here.

Also, the instruction on the value appears to be wrong based on the information under the example. Can the be lifted up here and removed from below?
```suggestion
`cdn_domains_list` is a map of the domain names that will be configured in CloudFront.
* the key is the fully qualified domain name
* the value is an array containing the internal prefix and the base domain (the application's Route 53 zone).  
```

File: README.md [hunk 5]
```diff
@@ -92,19 +92,52 @@ The R53 domains for non-production and production are stored in different AWS ac
...
+- Key: The endpoint name <myapp.mysite.com>
+- Values: application internal prefix and base domain <"internal", "mysite.com">
+
+### Optional settings:
+
+To create a R53 record pointing to the CloudFront endpoint set this to true.  If not set, in non production this is set to true by default and set to false in production.
+- enable_cdn_record: true
+
+To turn on CloudFront logging to a S3 bucket set this to true.
+- enable_logging: true
```
Comment (WillGibson) [inline README.md:138]: I would be inclined move this above the example, I just deleted the two questions I already asked there which this answers.

File: cdn/locals.tf [hunk 6]
```diff
@@ -0,0 +1,57 @@
+locals {
+  tags = {
+    application         = var.application
+    environment         = var.environment
+    managed-by          = "DBT Platform - Terraform"
+    copilot-application = var.application
+    copilot-environment = var.environment
+  }
+
+  # The primary domain for every application follows these naming standard.  See README.md
```
Comment (WillGibson) [inline cdn/locals.tf:10]: Wh
```suggestion
  # The primary domain for every application follows the naming standard documented under https://github.com/uktrade/terraform-platform-modules/blob/main/README.md#application-load-balancer-module
```

**Review by WillGibson — COMMENTED** [review]
Partial pass. Mostly comments on the documentation.

There's a lot to grok in the actual Terraform, but the tests feel kind of light by comparison.

File: README.md [hunk 7]
```diff
@@ -92,19 +92,52 @@ The R53 domains for non-production and production are stored in different AWS ac
...
-        dev.my-application.uktrade.digital: my-application.uktrade.digital
+        - dev.my-application.uktrade.digital: [ "internal", "my-application.uktrade.digital" ] 
+        - dev.my-web-service-2.my-application.uktrade.digital: [ "internal.my-web-service-2", "my-application.uktrade.digital" ]
       additional_address_list:
         - internal.my-web-service-2
+      enable_cdn_record: false
+      enable_logging: true
     prod:
-      domain: {my-application.great.gov.uk: "great.gov.uk"} 
+      cdn_domains_list: {my-application.great.gov.uk: "great.gov.uk"}
```
Comment (gabelton) [inline README.md:125]: Is it a problem that this isn't a list?

File: README.md [hunk 8]
```diff
@@ -92,19 +92,52 @@ The R53 domains for non-production and production are stored in different AWS ac
...
+      cdn_domains_list: {my-application.great.gov.uk: "great.gov.uk"} 
 ```
 
+Each item in the `cdn_domain_list` must include:
+- Key: The endpoint name <myapp.mysite.com>
+- Values: application internal prefix and base domain <"internal", "mysite.com">
+
+### Optional settings:
+
+To create a R53 record pointing to the CloudFront endpoint set this to true.  If not set, in non production this is set to true by default and set to false in production.
```
Comment (gabelton) [inline README.md:134]:
```suggestion
To create a R53 record pointing to the CloudFront endpoint, set this to true.  If not set, in non production this is set to true by default and set to false in production.
```

File: README.md [hunk 9]
```diff
@@ -92,19 +92,52 @@ The R53 domains for non-production and production are stored in different AWS ac
...
+- enable_cdn_record: true
+
+To turn on CloudFront logging to a S3 bucket set this to true.
```
Comment (gabelton) [inline README.md:137]:
```suggestion
To turn on CloudFront logging to a S3 bucket, set this to true.
```

File: .checkov.baseline (hunk 2 above)
Comment (ejayesh) [reply .checkov.baseline:31]: Gabe is sorting this one out

File: README.md (hunk 7 above)
Comment (ejayesh) [reply README.md:125]: Its a list of key: values eg.
{my-application.great.gov.uk: "great.gov.uk", my-secondapp.great.gov.uk: "great.gov.uk"}

File: README.md (hunk 5 above)
Comment (ejayesh) [reply README.md]: moved

File: .checkov.baseline (hunk 2 above)
Comment (ejayesh) [reply .checkov.baseline:31]: [DBTP-1039](https://uktrade.atlassian.net/browse/DBTP-1039)

File: README.md (hunk 7 above)
Comment (ejayesh) [reply README.md:135]: @gabelton is that ok, or am I missunderstanding your comment?

**Review by lgarvey — APPROVED** [review]
The terraform looks fine. Do we need some documentation?

File: README.md (hunk 7 above)
Comment (gabelton) [reply README.md:135]: but it's technically a dict / map? Should it not be in the same format as `cdn_domains_list` under dev above ?

**Comment by ejayesh** [comment]
> The terraform looks fine. Do we need some documentation?

I updated the readme and updated the platform docs

File: README.md (hunk 7 above)
Comment (ejayesh) [reply README.md:135]: oh my, you are correct.  I missed this, updaing.

File: cdn/main.tf [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+  type     = tolist(aws_acm_certificate.certificate[each.key].domain_validation_options)[0].resource_record_type
+  records  = [tolist(aws_acm_certificate.certificate[each.key].domain_validation_options)[0].resource_record_value]
+  ttl      = 300
+}
+
+resource "aws_cloudfront_distribution" "standard" {
+  # checkov:skip=CKV_AWS_305:This is managed in the application.
+  # checkov:skip=CKV_AWS_310:No fail-over origin required.
+  # checkov:skip=CKV2_AWS_32:Response headers policy not required.
+  # checkov:skip=CKV2_AWS_47:WAFv2 WebACL rules not set here.
```
Comment (WillGibson) [inline cdn/main.tf:47]: Can we say where they are set?

File: README.md (hunk 7 above)
Comment (ejayesh) [reply README.md]: updated

File: cdn/main.tf (hunk 10 above)
Comment (ejayesh) [reply cdn/main.tf]: updated

**Review by gabelton — APPROVED** [review]

**Review by WillGibson — APPROVED** [review]
For the bits I can meaningfully grok, mostly the documentation stuff, I'm adding my approval to @lgarvey's for the Terraform parts.

**Pull request merged**